var Base = require('ActionSet'),
	fn = require('functions');

// actions are executed in parallel, error means the end of execution
module.exports = Chaos = ActionSet.extend({

	// protected override
	runActions: function (args, callback) {
		var error = false,
			results = [],
			unfinished = this.actions.length,
			done = function (err) {
				if (error) {
					// do nothing, callback already called with error and partial results
				} else if (err) {
					error = true;
					callback(err, results);
				} else {
					results.push(Array.prototype.slice.call(arguments, 1));
					--unfinished;
					if (!unfinished) {
						callback(null, results);
					}
				}
			};

		if (unfinished) {
			this.actions.forEach(function (action, index) {
				action.apply(module, args.concat(this.arguments[index], [done]));
			}, this);
		} else {
			callback(null, []);
		}
	}

});